lib/src/sort.dart

```diff
@@ -0,0 +1,261 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// sort.dart
+// Implementation of sorting module.
+//
+// 2023 April 18
+// Author: Yao Jing Quek <yao.jing.quek@intel.com>
+//
+
+import 'dart:collection';
+
+import 'package:rohd/rohd.dart';
+import 'package:rohd_hcl/src/exceptions.dart';
+
+import 'package:rohd_hcl/src/utils.dart';
+
+/// An abstract class for all sort algorithm.
+abstract class Sort extends Module {
+  /// The List of logic to Sort
+  final Iterable<Logic> toSort;
+
+  /// Whether the sort [isAscending] order.
+  final bool isAscending;
+
+  /// Sort algorithm MUST have List of [toSort], direction of
+  /// sort [isAscending] and a name for the sorting module.
+  Sort({required this.toSort, this.isAscending = true, super.name});
+}
+
+/// Compare and Swap [Logic] to the specified order.
+class _CompareSwap extends Module {
+  final List<Logic> _inputs = [];
+
+  /// The list of sorted [Logic] result.
+  final List<Logic> _outputs = [];
+
+  /// The [swapped] list of result.
+  List<Logic> get swapped => _outputs;
+
+  /// The sorting [isAscending] that this [_CompareSwap] should
+  /// compare and swap.
+  final bool isAscending;
+
+  /// Compare and Swap the order of [i] and [j] in [toSort] based on the
+  /// direction given.
+  ///
+  /// The position [i] and [j] will swapped if [isAscending] is 1 and [_inputs]
+  /// of [i] greater than [_inputs] of [j] or [isAscending] is 0 and [_inputs]
+  /// of [i] lower than [_inputs] of [j]. The latency of the signals will only
+  /// be increased by 1 cycle if the signals for [i] and [j] are considered for
+  /// swapping.
+  _CompareSwap(Logic clk, Logic reset, List<Logic> toSort, int i, int j,
+      {required this.isAscending})
+      : super(name: 'compare_swap_${i}_$j') {
+    clk = addInput('clk', clk);
+    reset = addInput('reset', reset);
+
+    for (var i = 0; i < toSort.length; i++) {
+      _inputs.add(addInput('toSort$i', toSort[i], width: toSort[i].width));
+    }
+
+    final ascending = isAscending == true ? Const(1) : Const(0);
+    final newValA = Logic(width: toSort[0].width);
+    final newValB = Logic(width: toSort[0].width);
+
+    Sequential(clk, [
+      If(
+          (ascending & (_inputs[i] > _inputs[j])) |
```

Comment:
we can know `isAscending` at generation time, so we should just use a dart ternary operator here instead of using logic with `Const`s.

Reply:
I meant something like this:
```dart
      If( ascending ? (_inputs[i] > _inputs[j]) : _inputs[i].lt(_inputs[j])
```
Note how with this approach the generated hardware doesn't even consider whether it's ascending or descending, it just gets generated one way or the other.